realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1092,15 +1091,23 @@ public void run() {
                     if (handler != null
                             && !Thread.currentThread().isInterrupted()
                             && handler.getLooper().getThread().isAlive()) {
-                        if (onSuccess != null && transactionCommitted) {
+
+                        if (transactionCommitted) {
+                            // This will be treated like a special REALM_CHANGED event
                             handler.post(new Runnable() {
                                 @Override
                                 public void run() {
-                                    onSuccess.onSuccess();
+                                    handlerController.handleAsyncTransactionCompleted(onSuccess != null ? new Runnable() {
+                                        @Override
+                                        public void run() {
+                                            onSuccess.onSuccess();
+                                        }
+                                    } : null);
```

Comment:
How about just go for the normal routine `REALM_CHANGED` if the `onSuccess` is `null`? I am fine with this as well.